### ISSUE TITLE
bmv2 backend does not correctly cast bool to bit<1> in if statement

### ISSUE THREAD

**Issue by hanw** [p4lang/p4c#750]
Compiler bug reported by Rui Miguel,

simple_switch seems to misbehave when interpreting the bool type. I've used option -v to dump passes, as Mihai taught me, and it doesn't seem to be compiler-related, as the expressions in question are translated exactly as written by the final backend pass.

I can write the following:

### 1
```
m.metafield = true;

if (m.metafield) {
  //something happens
}
```
and it works just fine.

### 2

However, if I move the statement m.metafield = true; to the parser, the program produces awkward behavior.
```
parser TopParser(packet_in b, out Parsed_packet p, inout Meta m,
    inout standard_metadata_t metadata) {

  state start {
    m.metafield = true;
    transition accept;
  }
}

control IngressP(inout Parsed_packet hdr,
                      inout Meta m,
                      inout standard_metadata_t standard_metadata) {
apply {
  if (m.metafield) {
    //something happens
  }
}
```
An excerpt of the output switch log:
```
[11:24:12.581] [bmv2] [D] [thread 4689] [38.0] [cxt 0] Parser set: setting field 'scalars.Meta.metafield' from expression, new value is 1
[11:24:12.581] [bmv2] [T] [thread 4689] [38.0] [cxt 0] skeleton.p4(65) Condition "!m.metafieldfield" is true
```
### 3 

If I replace if (!m.metafield) for if (m.metafield == false), the switch doesn't even start, and the log shows:
```
lt-simple_switch: P4Objects.cpp:293: void bm::P4Objects::build_expression(const Json::Value&, bm::Expression*, bm::P4Objects::ExprType*): Assertion `typeL == typeR' failed.
```

Complete P4 program

```
#include <v1model.p4>

//------------------------------------------
header fixedLenHeader_h {
   bit<64> field;
}

//------------------------------------------
struct Meta {
   bool metafield;
}

struct Parsed_packet {
   fixedLenHeader_h h;
}

//------------------------------------------
parser TopParser(packet_in b, out Parsed_packet p, inout Meta m,
    inout standard_metadata_t metadata) {

  state start {
    m.metafield = true;
    transition accept;
  }
}

control VeryChecksum(in Parsed_packet hdr, inout Meta meta) { apply {} }

control IngressP(inout Parsed_packet hdr,
                      inout Meta m,
                      inout standard_metadata_t standard_metadata) {

  apply {
    if (m.metafield == false) {
      hdr.h.field = 64w5;
    }
  }
}

control EgressP(inout Parsed_packet hdr,
                     inout Meta meta,
                     inout standard_metadata_t standard_metadata) { apply {} }

control ChecksumComputer(inout Parsed_packet hdr,
                              inout Meta meta) { apply {} }

control TopDeparser(packet_out b, in Parsed_packet hdr) { apply {} }

V1Switch(TopParser(), VeryChecksum(), IngressP(), EgressP(), ChecksumComputer(),
    TopDeparser()) main;
```

**Comment by sethfowler** [comment]
This is the same problem that was reported in p4lang/behavioral-model#399. It's also related to #737.

**Comment by hanw** [comment]
fixed by #751

**Comment by mbudiu-vmw** [comment]
@hanw , since you claim this is fixed I will close it.
